rpc/gnoi/gnoi_cert.proto

```diff
@@ -0,0 +1,378 @@
+// associated to that Certificate ID.
+//
+// An Unimplemented error will be returned if the target cannot generate a CSR
+// as per the request. In this case, the caller must generate its own key pair.
+message GenerateCSRResponse {
+  CSR csr = 1;
+}
+
+// LoadCertificateRequest instructs the target to store the given certificate.
+//
+// Case 1: Target Generated CSR and Key Pair.
+// If the target generated the CSR (and the public/private key pair) during the
+// GenerateCSR request, then the target must associate the certificate with the
+// certificate ID specified in the preceding GenerateCSR request.
+//
+// Case 2: Externally Generated Key Pair.
+// If the target can not generate a CSR, then the public/private key pair is
+// generated externally. In this case provide the target with the key pair,
+// and the certificate_id to be associated with the new certificate.
+//
+// If there is another ongoing Rotate/Install RPC with the same certificate_id,
+// the LoadCertificateRequest must fail.
+message LoadCertificateRequest {
+  // The certificate to be Loaded on the target.
+  Certificate certificate = 1;
+
+  // The key pair to be used with the certificate. This is provided in the event
+  // that the target cannot generate a CSR (and the corresponding public/private
+  // keys).
+  KeyPair key_pair = 2;
+
+  // Certificate Id of the above certificate. This is to be provided only when
+  // there is an externally generated key pair.
+  string certificate_id = 3;
+}
+
+// Response from target after Loading a Certificate.
+// If the target could not load the certificate, it must end the RPC stream with
+// a suitable RPC error about why the Certificate was not loaded.
+message LoadCertificateResponse {
+}
+
+// A Finalize message is sent to the target to confirm the Rotation of
+// the certificate and that the certificate should not be rolled back when
+// the RPC concludes. The certificate must be rolled back if the target returns
+// an error after receiving a Finalize message.
+message FinalizeRequest {
+}
+
+// The request to query all the certificates on the target.
+message GetCertificatesRequest {
+}
+
+// Response from the target about the certificates that exist on the target what
+// what is using them.
+message GetCertificatesResponse {
+  repeated CertificateInfo certificate_info = 1;
+}
+
+message CertificateInfo {
+  string certificate_id = 1;
+  Certificate certificate = 2;
+
+  // List of endpoints using this certificate.
+  repeated Endpoint endpoints = 3;
+
+  // System modification time when the certificate was installed/rotated in
+  // nanoseconds since epoch.
+  int64 modification_time = 4;
+}
+
+message RevokeCertificatesRequest {
+  // Certificates to revoke.
+  repeated string certificate_id = 1;
+}
+
+message RevokeCertificatesResponse {
+  // List of certificates successfully revoked.
+  repeated string revoked_certificate_id = 1;
+
+  // List of errors why certain certificates could not be revoked.
```

Comment:
Should this just be included as a `proto.Any` in `status.Status` [https://github.com/googleapis/googleapis/blob/master/google/rpc/status.proto]?